File: experiment/websites/js/ackermann/ackermann.js
/* based on: https://rosettacode.org/wiki/Ackermann_function#ES5 (2022-01-14) */

/**
 * @param {number} m
 * @param {number} n
 * @returns {number}
 */
function ack(m, n) {
  return m === 0 ? n + 1 : ack(m - 1, n === 0  ? 1 : ack(m, n - 1));
}

File: experiment/websites/c/ackermann/ackermann.c
/* based on: https://rosettacode.org/wiki/Ackermann_function#C (2022-01-14) */

#include <emscripten/emscripten.h>

int ackermann(int m, int n) {
  if (!m) return n + 1;
  if (!n) return ackermann(m - 1, 1);
  return ackermann(m - 1, ackermann(m, n - 1));
}

EMSCRIPTEN_KEEPALIVE
int run_ackermann(int x, int y) {
  for (int m = 0; m <= x; m++)
    for (int n = 0; n <= y; n++)
      ackermann(m, n);
  return 0;
}
